docs/architecture/adr-001-coin-source-tracing.md

```diff
@@ -0,0 +1,428 @@
+func (dt *DenomTrace) RemovePrefix() error {
+  if dt.Trace == "" {
+    return nil
+  }
+
+  traceSplit := strings.SplitN(dt.Trace, "/", 3)
+
+  var err error
+  switch {
+  case len(traceSplit) == 1 && traceSplit[0] == dt.Trace:
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must contain '/' separators", dt.Trace)
+  case len(traceSplit) == 1:
+    // trace only contains one identifier
+    err = Wrapf(ErrInvalidDenomForTransfer, "trace info %s must come in pairs of identifiers '{portID}/{channelID}'", dt.Trace)
+  case len(traceSplit) == 2:
+    dt.Trace = ""
+  case len(traceSplit) == 3:
+    dt.Trace = traceSplit[2]
+  }
+
+  if err != nil {
+    return err
+  }
+
+  return nil
+}
+```
+
+### `x/ibc-transfer` Changes
+
+In order to retrieve the trace information from an IBC denomination, a lookup table needs to be
+added to the `ibc-transfer` module. These values need to also be persisted between upgrades, meaning
+that a new `[]DenomTrace` `GenesisState` field state needs to be added to the module:
+
+```golang
+// GetDenom retrieves the full identifiers trace and base denomination from the store.
+func (k Keeper) GetDenomTrace(ctx Context, denomTraceHash []byte) (DenomTrace, bool) {
+  store := ctx.KVStore(k.storeKey)
+  bz := store.Get(types.KeyDenomTrace(traceHash))
+  if bz == nil {
+    return &DenomTrace, false
+  }
+
+  var denomTrace DenomTrace
+  k.cdc.MustUnmarshalBinaryBare(bz, &denomTrace)
+  return denomTrace, true
+}
+
+// HasTrace checks if a the key with the given trace hash exists on the store.
+func (k Keeper) HasDenomTrace(ctx Context, denomTraceHash []byte)  bool {
+  store := ctx.KVStore(k.storeKey)
+  return store.Has(types.KeyTrace(denomTraceHash))
+}
+
+// SetTrace sets a new {trace hash -> trace} pair to the store.
+func (k Keeper) SetTrace(ctx Context, denomTrace DenomTrace) {
+  store := ctx.KVStore(k.storeKey)
+  bz := k.cdc.MustMarshalBinaryBare(&denomTrace)
+  store.Set(types.KeyTrace(denomTrace.Hash()), bz)
+}
+```
+
+The problem with this approach is that when a token is received for the first time, the full trace
+info will need to be passed in order to construct the hash and set it to the mapping on the store.
+To mitigate this a new `DenomTrace` field needs to be added to the `FungibleTokenPacketData`:
+
+```protobuf
+message FungibleTokenPacketData {
+  // the token denomination to be transferred
+  string denom = 1;
+  // the token amount to be transferred
+  uint64 amount = 2;
+  // coins denomination trace for tracking the source
+  DenomTrace denom_trace = 3;
+  // the sender address
+  string sender = 4;
+  // the recipient address on the destination chain
+  string receiver = 5;
+}
+```
+
+The `MsgTransfer` will validate that the `Coin` denomination from the `Token` field contains a valid hash:
+
+```golang
+func (msg MsgTransfer) ValidateBasic() error {
```

Comment:
so I think `MsgTransfer` should take in the ibc/hash and then use the lookup table to construct the full trace and place that into the fungible token packet. Clients could allow the user to provide the hash or the display version and then use the lookup to provide the hash. 

this cuts down on validation needed here and makes the logic easier.

MsgTransfer takes in internally represented token -> passes to SendTransfer which lookups up the full denom and constructs the fungible token packet data for an IBC transfer

Reply:
`MsgTransfer` can take either the "ibc/hash" denom or the base denomination. Maybe it's worth clarifying that it won't accept the 'portID/channelID/denom' tokens. 

> so I think MsgTransfer should take in the ibc/hash and then use the lookup table to construct the full trace and place that into the fungible token packet. 

yeah for the first case that's exactly what it does.

I'd leave the display version of the bank denom metadata out of this ADR as it is out of scope.